### PR TITLE
Simplify `ChangeSet` and introduce `Additions`.

### PR DESCRIPTION
**Pull request by evanlinjin**
This is a portion of the work pulled from #64

### CODE REVIEW

File: bdk_core/src/sparse_chain.rs [hunk 1]
```diff
@@ -501,8 +506,8 @@ impl<I: ChainIndex> SparseChain<I> {
 /// Represents the set of changes as result of a successful [`Update`].
 #[derive(Debug, PartialEq)]
 pub struct ChangeSet<I = TxHeight> {
-    pub checkpoints: HashMap<u32, Change<BlockHash>>,
-    pub txids: HashMap<Txid, Change<I>>,
+    pub checkpoints: HashMap<u32, Option<BlockHash>>,
+    pub txids: HashMap<Txid, Option<I>>,
```
Comment (rajarshimaitra) [inline bdk_core/src/sparse_chain.rs:510]: Not a review but a conceptual question. Now that we are removing the `Change` structure, and from the test code it seems the `Option` represents the final state of the change, Is that correct? Also won't it be a useful verbosity to keep the "full change", like "From this" - "To this" pattern? Or is the thought is that can always implicitly figure out what changed from what with the final state only?

Comment (LLFourn) [reply bdk_core/src/sparse_chain.rs:510]: Yes we can figure it out just by looking at the `SparseChain` and the changeset.

File: bdk_core/src/chain_graph.rs [hunk 2]
```diff
@@ -38,7 +49,7 @@ impl<I: ChainIndex> ChainGraph<I> {
     pub fn insert_output(
         &mut self,
         outpoint: OutPoint,
-        txout: TxOut,
+        txout: &TxOut,
```
[the remark this answers is not in the document]
Comment (evanlinjin) [reply bdk_core/src/chain_graph.rs:52]: True, the way I was thinking about it is that if we had a reference, we can avoid additional cloning if txout, or a full tx already existed.

File: bdk_core/src/chain_graph.rs [hunk 3]
```diff
@@ -57,26 +68,70 @@ impl<I: ChainIndex> ChainGraph<I> {
...
+    pub fn determine_changeset(&self, update: &Self) -> Result<ChangeSet<I>, UpdateFailure<I>> {
+        let chain_changeset = self.chain.determine_changeset(&update.chain)?;
+        let graph_additions = self.graph.determine_additions(&update.graph);
+
+        let added_txids = graph_additions.txids::<HashSet<_>>();
+
+        // ensure changeset adds exist in either graph_additions or self.graph
+        let missing = chain_changeset
+            .tx_additions()
+            .filter(|txid| added_txids.contains(txid) || self.graph.contains_txid(*txid))
```
[the remark this answers is not in the document]
Comment (evanlinjin) [reply bdk_core/src/chain_graph.rs:80]: Via the DMs, we have decided to not have such a check enforced in `determine_changeset`. Instead, we include helper methods such as `missing_txs()` if the user wishes for such checking.

**Pull request merged**